### PR TITLE
Fix missed content classes after menu item re-rendering (T586310)

### PR DESCRIPTION
**Pull request by sometimesElen**

### CODE REVIEW

File: js/ui/context_menu/ui.menu_base.js [hunk 1]
```diff
@@ -633,20 +631,57 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
...
+                $nodeElement = $item.closest("li");
+            this._renderItem(index, node, $nodeContainer, $nodeElement);
+        }
+    },
+
     _addContentClasses: function(node, $itemFrame) {
         if(this._displayGetter(node)) {
             $itemFrame.addClass(DX_ITEM_HAS_TEXT);
+        } else {
+            $itemFrame.removeClass(DX_ITEM_HAS_TEXT);
```
Comment (DokaRus) [inline js/ui/context_menu/ui.menu_base.js]: use the `toggleClass` method

File: js/ui/context_menu/ui.menu_base.js [hunk 2]
```diff
@@ -633,20 +631,57 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
...
         }
 
         // deprecated since 15.1 (itemData.iconSrc)
         if(node.icon || node.iconSrc) {
             $itemFrame.addClass(DX_ITEM_HAS_ICON);
             this.hasIcons = true;
+        } else {
+            $itemFrame.removeClass(DX_ITEM_HAS_ICON);
```
Comment (DokaRus) [inline js/ui/context_menu/ui.menu_base.js]: you can refactor it too

File: js/ui/context_menu/ui.menu_base.js [hunk 3]
```diff
@@ -633,20 +631,57 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
...
         }
 
-        if(this._hasSubmenu(node)) {
+        if(this._hasSubmenu(node) || (this._itemsGetter(node) && this._itemsGetter(node).length)) {
             $itemFrame.addClass(DX_ITEM_HAS_SUBMENU);
+        } else {
+            $itemFrame.removeClass(DX_ITEM_HAS_SUBMENU);
```
Comment (DokaRus) [inline js/ui/context_menu/ui.menu_base.js]: use the `toggleClass` method

File: js/ui/context_menu/ui.menu_base.js [hunk 4]
```diff
@@ -633,20 +631,57 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
         if(this._hasSubmenu(node)) this.setAria("haspopup", "true", $itemFrame);
     },
 
+    _renderItemFrame: function(index, itemData, $container) {
+        var $itemFrame = $container.children(".dx-menu-item");
+        if(!$itemFrame || !$itemFrame.length) {
```
Comment (DokaRus) [inline js/ui/context_menu/ui.menu_base.js]: Is there a chance, that the `children` function returns `null` or `undefined`?

File: js/ui/context_menu/ui.menu_base.js [hunk 5]
```diff
@@ -633,20 +631,57 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
         if(this._hasSubmenu(node)) this.setAria("haspopup", "true", $itemFrame);
     },
 
+    _renderItemFrame: function(index, itemData, $container) {
+        var $itemFrame = $container.children(".dx-menu-item");
```
Comment (DokaRus) [inline js/ui/context_menu/ui.menu_base.js]: extract a class names to constant

**Review by DokaRus — CHANGES_REQUESTED** [review]

**Comment by DokaRus** [comment]
Let's wait for @mrjohn0011 review

File: js/ui/context_menu/ui.menu_base.js [hunk 6]
```diff
@@ -566,7 +566,7 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
     },
 
     _hasSubmenu: function(node) {
-        return node.internalFields.childrenKeys.length;
+        return node.internalFields && node.internalFields.childrenKeys.length;
```
Comment (mrjohn0011) [inline js/ui/context_menu/ui.menu_base.js]: If a node has no internalFields - it is an item. Why not converted items appear in this method? This issue should be fixed on another level

File: js/ui/context_menu/ui.menu_base.js [hunk 7]
```diff
@@ -626,27 +624,53 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
...
+        if(!this.constructor.ItemClass.getInstance($item).setDataField(property, value)) {
+            var node = this._dataAdapter.getNodeByItem(item),
+                index = $item.data(this._itemIndexKey()),
+                $nodeContainer = $item.closest("ul"),
+                $nodeElement = $item.closest("li");
+            this._renderItem(index, node, $nodeContainer, $nodeElement);
+        }
+    },
+
+    _addContentClasses: function(node, $itemFrame) {
```
Comment (mrjohn0011) [inline js/ui/context_menu/ui.menu_base.js]: If this is not a `node`, rename please

File: js/ui/context_menu/ui.menu_base.js [hunk 8]
```diff
@@ -626,27 +624,53 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
...
+
+        return $itemFrame.length ? $itemFrame : this.callBase.apply(this, arguments);
+    },
+
+    _itemOptionChanged: function(item, property, value) {
+        var $item = this._findItemElementByItem(item);
+        if(!$item.length) {
+            return;
         }
+        if(!this.constructor.ItemClass.getInstance($item).setDataField(property, value)) {
```
Comment (mrjohn0011) [inline js/ui/context_menu/ui.menu_base.js]: Don't redefine all method but only required part of this

File: testing/tests/DevExpress.ui.widgets/menuBase.tests.js [hunk 9]
```diff
@@ -258,6 +259,36 @@ QUnit.test("item container should not have dx-menu-no-icons class when at least
...
+    //arrange
+    var menuBase = createMenu({ items: [{ text: "item" }] }),
+        $item;
+
+    //act
+    menuBase.instance.option("items[0].icon", "add");
+    $item = menuBase.element.find('.' + DX_MENU_ITEM_WRAPPER_CLASS).children();
+
+    //assert
+    assert.equal($item.length, 1, 'there is 1 item in menu');
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets/menuBase.tests.js]: This assert is redundant

**Review by mrjohn0011 — CHANGES_REQUESTED** [review]
Pull request should unclear name. The name should be corrected to describe fixed problem more concrete

File: js/ui/collection/ui.collection_widget.base.js [hunk 10]
```diff
@@ -505,12 +505,16 @@ var CollectionWidget = Widget.inherit({
             return;
         }
         if(!this.constructor.ItemClass.getInstance($item).setDataField(property, value)) {
-            var itemData = this._getItemData($item),
-                index = $item.data(this._itemIndexKey());
-            this._renderItem(index, itemData, null, $item);
+            this._renderNewItem($item, item);
```
Comment (DokaRus) [inline js/ui/collection/ui.collection_widget.base.js]: You update an existed item, right? Maybe better rename to `_refreshItem`?

File: js/ui/context_menu/ui.menu_base.js [hunk 11]
```diff
@@ -626,27 +624,47 @@ var MenuBase = HierarchicalCollectionWidget.inherit({
...
+
+        $itemFrame.toggleClass(DX_ITEM_HAS_SUBMENU, hasSubmenu);
+    },
 
-        if(this._hasSubmenu(node)) {
-            $itemFrame.addClass(DX_ITEM_HAS_SUBMENU);
+    _getItemContent: function($itemFrame) {
+        var $itemContent = this.callBase($itemFrame);
+        if(!$itemContent.length) {
+            $itemContent = $itemFrame.children("." + ITEM_CLASS + "-content");
```
Comment (DokaRus) [inline js/ui/context_menu/ui.menu_base.js]: Move to a constant, do not hide classes or selectors in the code

**Review by DokaRus — APPROVED** [review]

**Review by mrjohn0011 — APPROVED** [review]

**Pull request merged**